### PR TITLE
AdonisJS 4 and laravel compatible bcrypt

### PR DESCRIPTION
**Pull request by reg2005**
## Proposed changes

Add legacy Bcrypt driver, [discussion here](https://github.com/adonisjs/core/discussions/1549#discussioncomment-59994)

## Types of changes

What types of changes does your code introduce?

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/adonisjs/hash/blob/master/CONTRIBUTING.md) doc
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: src/Drivers/BcryptLegacy.ts [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+				if (error) {
+					return reject(error)
+				}
+				resolve(hash)
+			})
+		})
+	}
+
+	public needsReHash(hashedValue: string): boolean {
+		if (hashedValue.startsWith('$2y')) {
```
Comment (RodolfoSilva) [inline src/Drivers/BcryptLegacy.ts:60]: This does not cover the other `bcrypt` formats like: `$2a$ `and `$2b$`, maybe you should apply a simple regex for that
```ts
/^\$2[aby]/.test(hashedValue)
```

File: src/Drivers/BcryptLegacy.ts [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+ * and
+ * (c) Evgeniy Ryumin <cmp08@ya.ru>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+/// <reference path="../../adonis-typings/hash.ts" />
+import { BcryptLegacyContract, BcryptLegacyConfig } from '@ioc:Adonis/Core/Hash'
+import bcrypt from 'bcryptjs'
```
Comment (RodolfoSilva) [inline src/Drivers/BcryptLegacy.ts:14]: Why not use the `bcrypt` library instead `bcryptjs`?

**Comment by thetutlage** [comment]
Instead of creating a new driver. I was expecting to have a boolean flag in the `bcrypt` config.

```ts
{
   driver: 'bcrypt',
   legacy: true
}
```

What you think?

**Comment by RodolfoSilva** [comment]
@thetutlage I agree! It would be great. 😁

**Comment by reg2005** [comment]
> Instead of creating a new driver. I was expecting to have a boolean flag in the `bcrypt` config.
> 
> ```ts
> {
>    driver: 'bcrypt',
>    legacy: true
> }
> ```
> 
> What you think?

I think when developer uses bcryptLegacy driver, he understands what he is doing, because he will need install bcryptjs for right work. In future when he will to move to standard bcrypt or argon, he just will remove bcryptjs from dependencies and switch default hash driver to argon or main bcrypt.

Current version of my PR is explicit and simple and also follow of original architecture ideas this repo

Maybe legacyMode will bring in some entanglement.

File: src/Drivers/BcryptLegacy.ts (hunk 2 above)
Comment (DraftProducts) [reply src/Drivers/BcryptLegacy.ts:14]: I saw that Adonis V4 use `bcrypt` package, do the two hashes have the same format?

**Comment by DraftProducts** [comment]
I prefer the boolean legacy too. 
If you want i can make the necessary modifications :)

**Comment by thetutlage** [comment]
@DraftProducts Please happy to accept the PR :)

**Comment by thetutlage** [comment]
Closing as not actionable. Feel free to create a new PR as per the described expectations

**Pull request closed without merging**